cmd/clusterctl/client/cluster/template.go

```diff
@@ -181,35 +182,81 @@ func (t *templateClient) getGitHubFileContent(rURL *url.URL) ([]byte, error) {
 
 	// Extract all the info from url split.
 	owner := urlSplit[0]
-	repository := urlSplit[1]
+	repo := urlSplit[1]
+	linkType := urlSplit[2]
 	branch := urlSplit[3]
 	path := strings.Join(urlSplit[4:], "/")
 
 	// gets the GitHub client
-	client, err := t.gitHubClientFactory(t.configClient.Variables())
+	ghClient, err := t.gitHubClientFactory(t.configClient.Variables())
 	if err != nil {
 		return nil, err
 	}
 
 	// gets the file from GiHub
-	fileContent, _, _, err := client.Repositories.GetContents(context.TODO(), owner, repository, path, &github.RepositoryContentGetOptions{Ref: branch})
+	switch linkType {
+	case "blob": // get file from a code in a github repo
+		return getGithubFileContentFromCode(ghClient, rURL.Path, owner, repo, path, branch)
+
+	case "releases": // get a github release asset
+		tag := urlSplit[4]
+		assetName := urlSplit[5]
+
+		return getGithubAssetFromRelease(ghClient, rURL.Path, owner, repo, tag, assetName)
+
+	default:
+		return nil, fmt.Errorf("unknown github URL: %v", rURL)
+	}
+}
+
+func getGithubFileContentFromCode(ghClient *github.Client, fullPath string, owner string, repo string, path string, branch string) ([]byte, error) {
+	fileContent, _, _, err := ghClient.Repositories.GetContents(ctx, owner, repo, path, &github.RepositoryContentGetOptions{Ref: branch})
 	if err != nil {
-		return nil, handleGithubErr(err, "failed to get %q", rURL.Path)
+		return nil, handleGithubErr(err, "failed to get %q", fullPath)
 	}
 	if fileContent == nil {
-		return nil, errors.Errorf("%q does not return a valid file content", rURL.Path)
+		return nil, errors.Errorf("%q does not return a valid file content", fullPath)
 	}
 	if fileContent.Encoding == nil || *fileContent.Encoding != "base64" {
-		return nil, errors.Errorf("invalid encoding detected for %q. Only base64 encoding supported", rURL.Path)
+		return nil, errors.Errorf("invalid encoding detected for %q. Only base64 encoding supported", fullPath)
 	}
-
 	content, err := base64.StdEncoding.DecodeString(*fileContent.Content)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to decode file %q", rURL.Path)
+		return nil, errors.Wrapf(err, "failed to decode file %q", fullPath)
 	}
 	return content, nil
 }
 
+func getGithubAssetFromRelease(ghClient *github.Client, path string, owner string, repo string, tag string, assetName string) ([]byte, error) {
+	release, _, err := ghClient.Repositories.GetReleaseByTag(ctx, owner, repo, tag)
+	if err != nil {
+		return nil, handleGithubErr(err, "failed to get release '%s' from %s/%s repository", tag, owner, repo)
+	}
+
+	if release == nil {
+		return nil, fmt.Errorf("can't find release '%s' in %s/%s repository", tag, owner, repo)
+	}
+
+	var rc io.ReadCloser
+	for _, asset := range release.Assets {
+		if asset.GetName() == assetName {
+			rc, _, err = ghClient.Repositories.DownloadReleaseAsset(ctx, owner, repo, asset.GetID(), ghClient.Client())
```

Comment:
I believe you will also have to handle the redirects when working with release assets. Similar to how we handle the redirects here: 
https://github.com/nunnatsa/cluster-api/blob/fd24c247c3ae9b32fed236f544088486ec6331cf/cmd/clusterctl/client/repository/repository_github.go#L331-L344

Reply:
> I believe you will also have to handle the redirects when working with release assets. Similar to how we handle the redirects here: https://github.com/nunnatsa/cluster-api/blob/fd24c247c3ae9b32fed236f544088486ec6331cf/cmd/clusterctl/client/repository/repository_github.go#L331-L344

Thanks @ykakarap. I'm not sure it's needed. According to the  `DownloadReleaseAsset` documentation:
> followRedirectsClient can be passed to download the asset from a redirected location. Passing http.DefaultClient is recommended unless special circumstances exist, but it's possible to pass any http.Client. **If nil is passed the redirectURL will be returned instead**.

I added unit test to check the redirect scenario. Also, This code is working with the real github, that actually responses with redirect. If you think it's a must, I'll add it, but it's not needed, I think.

Reply:
Seem to work for redirected assets as well. @fabriziopandini Do you know why we are redirecting here: https://github.com/nunnatsa/cluster-api/blob/fd24c247c3ae9b32fed236f544088486ec6331cf/cmd/clusterctl/client/repository/repository_github.go#L331-L344?

Do you know of a case that is not covered by the standard `DownloadReleaseAsset` function?

Reply:
Thanks @fabriziopandini. According to the `DownloadReleaseAsset` documentation I copied above, the redirect is only returned if the followRedirectsClient parameter is empty. This is, I guess, if you want to implement the redirect logic. But if you pass the client, the function will do it for you. This is how I understand it.

Reply:
kk, I double checked the DownloadReleaseAsset implementation and it seems that this is not necessary if we are passing a followRedirectsClient.
In a separated PR we can also eventually clean up also the bit of code @ykakarap was referencing to

Reply:
Created a follow-up issue to track the clean up work. https://github.com/kubernetes-sigs/cluster-api/issues/7942